pkg/resource/pod-resource-manger.go

```diff
@@ -0,0 +1,204 @@
+	if err != nil {
+		klog.Errorf("GetRuntimeClient failed %s", err.Error())
+		return nil
+	}
+
+	o := &PodResourceManager{
+		nodeName:      nodeName,
+		client:        client,
+		podLister:     podInformer.Lister(),
+		podSynced:     podInformer.Informer().HasSynced,
+		runtimeClient: runtimeClient,
+		runtimeConn:   runtimeConn,
+		stateChann:    stateChann,
+		collectors:    Collectors,
+	}
+	podInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		// Focused on pod belonged to this node
+		FilterFunc: func(obj interface{}) bool {
+			switch t := obj.(type) {
+			case *v1.Pod:
+				return ownedPod(t, o.nodeName)
+			default:
+				utilruntime.HandleError(fmt.Errorf("unable to handle object %T", obj))
+				return false
+			}
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: o.reconcilePod,
+			UpdateFunc: func(old, cur interface{}) {
+				o.reconcilePod(cur)
+			},
+			DeleteFunc: o.reconcilePod,
+		},
+	})
+	return o
+}
+
+func (o *PodResourceManager) Name() string {
+	return "PodResourceManager"
+}
+
+func (o *PodResourceManager) Run(stop <-chan struct{}) {
+	klog.Infof("Starting pod resource manager.")
+
+	// Wait for the caches to be synced before starting workers
+	if !cache.WaitForNamedCacheSync("pod-resource-manager",
+		stop,
+		o.podSynced,
+	) {
+		return
+	}
+
+	go func() {
+		for {
+			select {
+			case state := <-o.stateChann:
+				o.state = state
+				o.lastStateTime = time.Now()
+			case <-stop:
+				klog.Infof("pod resource manager exit")
+				if err := cgrpc.CloseGrpcConnection(o.runtimeConn); err != nil {
+					klog.Errorf("Failed to close grpc connection: %v", err)
+				}
+				return
+			}
+		}
+	}()
+
+	return
+}
+
+func (o *PodResourceManager) reconcilePod(obj interface{}) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("Couldn't get object from tombstone %#v", obj))
+			return
+		}
+		pod, ok = tombstone.Obj.(*v1.Pod)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("Tombstone contained object that is not a pod %#v", obj))
+			return
+		}
+	}
+
+	o.updatePodExtResToCgroup(pod)
+}
+
+func ownedPod(pod *v1.Pod, nodeName string) bool {
+	return pod.Spec.NodeName == nodeName && pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed
+}
+
+// Get pod's ext-resource.node.gocrane.io resource and update it to Cgroup
+func (o *PodResourceManager) updatePodExtResToCgroup(pod *v1.Pod) {
+	for _, c := range pod.Spec.Containers {
+		for res, val := range c.Resources.Limits {
+			if strings.HasPrefix(res.String(), fmt.Sprintf(ExtResourcePrefixFormat, v1.ResourceCPU)) {
+				containerId := getContainerIdFromPod(pod, c.Name)
+				if containerId == "" {
+					continue
+				}
+
+				containerPeriod := o.getCPUPeriod(pod, containerId)
+				if containerPeriod == 0 {
+					continue
+				}
+
+				// Update cpu quota by CRI
```

Comment:
should you compare quota before update?
we should avoid cgroup change on every pod update event.

Reply:
I discussed with devin, pod resource manager only convert containers whose quota is -1, if it has value, the adjustment can handled by throttle module. And throttle module has the logic about compare.